lib/src/rules/avoid_dynamic_calls.dart

```diff
@@ -0,0 +1,269 @@
+  }
+}
+
+class _Visitor extends SimpleAstVisitor<void> {
+  final LintRule rule;
+
+  _Visitor(this.rule);
+
+  bool _lintIfDynamic(Expression node) {
+    if (node?.staticType?.isDynamic == true) {
+      rule.reportLint(node);
+      return true;
+    } else {
+      return false;
+    }
+  }
+
+  void _lintIfDynamicOrFunction(Expression node) {
+    final staticType = node?.staticType;
+    if (staticType == null) {
+      return;
+    }
+    if (staticType.isDynamic) {
+      rule.reportLint(node);
+    }
+    if (staticType.isDartCoreFunction) {
+      rule.reportLint(node);
+    }
+  }
+
+  @override
+  void visitAssignmentExpression(AssignmentExpression node) {
+    if (node.readType?.isDynamic != true) {
+      // An assignment expression can only be a dynamic call if it is a
+      // "compound assignment" (i.e. such as `x += 1`); so if `readType` is not
+      // dynamic, we don't need to check further.
+      return;
+    }
+    if (node.operator.type == TokenType.QUESTION_QUESTION_EQ) {
+      // x ??= foo is not a dynamic call.
+      return;
+    }
+    rule.reportLint(node);
+  }
+
+  @override
+  void visitBinaryExpression(BinaryExpression node) {
+    if (!node.operator.isUserDefinableOperator) {
+      // Operators that can never be provided by the user can't be dynamic.
+      return;
+    }
+    switch (node.operator.type) {
+      case TokenType.EQ_EQ:
+      case TokenType.BANG_EQ:
+        // These operators exist on every type, even "Object?". While they are
+        // virtually dispatched, they are not considered dynamic calls by the
+        // CFE. They would also make landing this lint exponentially harder.
+        return;
+    }
+    _lintIfDynamic(node.leftOperand);
+    // We don't check node.rightOperand, because that is an implicit cast, not a
+    // dynamic call (the call itself is based on leftOperand). While it would be
+    // useful to do so, it is better solved by other more specific lints to
+    // disallow implicit casts from dynamic.
+  }
+
+  @override
+  void visitFunctionExpressionInvocation(FunctionExpressionInvocation node) {
+    if (node.function != null) {
+      _lintIfDynamicOrFunction(node.function);
+    }
+  }
+
+  @override
+  void visitIndexExpression(IndexExpression node) {
+    _lintIfDynamic(node.realTarget);
+  }
+
+  @override
+  void visitMethodInvocation(MethodInvocation node) {
+    final methodName = node.methodName.name;
+    if (methodName == 'noSuchMethod' || methodName == 'toString') {
+      // Special-cased; these exist on every object, even those typed "Object?".
+      return;
+    }
+    final receiverWasDynamic = _lintIfDynamic(node.realTarget);
+    if (!receiverWasDynamic && node.target == null) {
+      _lintIfDynamicOrFunction(node.function);
+    }
+  }
+
+  @override
+  void visitPrefixExpression(PrefixExpression node) {
+    // Implemented as: node?.staticType?.isDynamic == true
+    if (_lintIfDynamic(node.operand)) {
```

Comment:
Thank you for the clarification.

I think we want to to such switch on the operator to explicitly express what we are checking and when. Doing one generic `_lintIfDynamic` when for `++x` the type is `null` does not make this code as easy to read as it could be.

And for `++` and `--` we want to check `CompoundAssignmentExpression.readType` directly. Because this is the type on which we will call `operator+` or `operator-`. Just like we do for assignments.

Similarly for postfix expressions.

Reply:
Thanks, that's a solid idea. I've consolidated and de-duplicated the code into `_lintPrefixOrPostfixExpression(node, node.operand);`. PTAL.